### PR TITLE
Add support for pipelines providers defaults

### PR DESCRIPTION
**Pull request by rporres**
covers [APPSRE-4165](https://issues.redhat.com/browse/APPSRE-4165)

related schema changes: https://github.com/app-sre/qontract-schemas/pull/44

Signed-off-by: Rafa Porres Molina <rporresm@redhat.com>

### CODE REVIEW

**Review by steveteahan — COMMENTED** [review]
Thanks for moving the tests to pytest! I'm not as familiar with this part of the code base, so I'll allow someone else with more knowledge of the change to approve.

File: reconcile/test/test_queries.py [hunk 1]
```diff
@@ -1,25 +1,63 @@
...
+
+    def teardown_method(self) -> None:
+        """Cleanup patches created in self.setup_method"""
+        self.gql_patcher.stop()
+
+    def mock_gql_query(self, query: str) -> dict[str, Any]:
+        return self.fixture_data
+
+    def test_get_permissions_return_all_slack_usergroup(self) -> None:
+        self.fixture_data = Fixtures('slack_usergroups'). \
```
[the remark this answers is not in the document]
Comment (rporres) [reply reconcile/test/test_queries.py:27]: Thx a lot for the long comment, very insightful.

I agree that `self.fixture_data` is a bit obscure. I tried several ways to set it in the `setup_method` but I had a few problems with mypy so I didn't declare it in the setup method, which makes things a bit magic.

I wanted a solution that avoided repeating the same code for every test and worked with any query without touching the existing fixtures and that could easily work with new tests without changing too much the code. I'll take a look into ways to make it clearer to the reader.

[the remark this answers is not in the document]
Comment (rporres) [reply reconcile/test/test_queries.py:33]: Done in https://github.com/app-sre/qontract-reconcile/pull/2080/commits/99eb6069a3c0924fc8183315cd78fe3287ef902f

File: reconcile/test/fixtures/queries/pipelines_providers_no_defaults.json [hunk 2]
```diff
@@ -0,0 +1,60 @@
+{
```
[the remark this answers is not in the document]
Comment (rporres) [reply reconcile/test/fixtures/queries/pipelines_providers_no_defaults.json:1]: My xunit/patch background shows... Thanks for the pointers :)

**Comment by rporres** [comment]
Note for the reviewer: the above comments from @steveteahan regarding patch vs creation methods and the dangers of DRY are really worth the read. Resolving them since they're in a wider scope of this PR but they are quite interesting.

**Review by Piojo — COMMENTED** [review]
LGTM. Thanks for the testing. :)

**Comment by rporres** [comment]
> Does it make sense to set defaults for taskTemplates and pipelineTemplates ?

Yes. Every provider is currently using the same task and pipeline templates, so it's very valuable to extract them to a default. In the past some people expressed interest in tailoring the pipeline for their needs. While we tried to avoid it being creative, we finally saw that there was no way to get certain things. But it is not the general case, so the overridable defaults model for tasks and pipelines makes sense

**Comment by rporres** [comment]
> Checking the data directory I see there is a lot of duplication on taskTemplates and pipelineTemplates. Maybe is worth to consider extracting this data to its own object and reference it in pipeline-provider? IT seems that the majority of pipeline providers are just changing the namespace reference.

I don't understand well this.

**Comment by lechuk47** [comment]
> > Checking the data directory I see there is a lot of duplication on taskTemplates and pipelineTemplates. Maybe is worth to consider extracting this data to its own object and reference it in pipeline-provider? IT seems that the majority of pipeline providers are just changing the namespace reference.
> 
> I don't understand well this.

let me try to clarify it: 

As you said, almost all pipeline providers are using the same templates, so I'm wondering if it makes sense to have some kind of "pipeline" object and reference it in all the pipeline providers. I'm wondering if it makes sense to have it as an object, not as a templates default value. 

If I'm not wrong, `pipeline-providers` should cover other pipeline implementations other than `openshift-saas-deploy`, so setting it as a default value does not seem a good approach to me.

**Comment by rporres** [comment]
> If I'm not wrong, pipeline-providers should cover other pipeline implementations other than openshift-saas-deploy, so setting it as a default value does not seem a good approach to me.

Since we can only trigger pipelineruns for openshift-saas-deploy, I made it quite obvious everywhere that we could only create that kind of pipelines. So in fact, the schema is a bit deceiving in being so general... And it can be confusing. I'll add a comment on it.

**Comment by rporres** [comment]
@lechuk47 Added clarification on `pipelineTemplate` in https://github.com/app-sre/qontract-schemas/pull/48

**Comment by rporres** [comment]
b

**Pull request merged**